sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -65,35 +92,96 @@ class EquivalentExpressions {
   }
 
   /**
-   * Adds the expression to this data structure recursively. Stops if a matching expression
-   * is found. That is, if `expr` has already been added, its children are not added.
+   * Adds each expression to structural expression data structure, grouping them with existing
+   * structurally equivalent expressions. Non-recursive. Returns false if this doesn't add input
+   * expression actually.
+   */
+  def addStructExpr(expr: Expression): Boolean = {
+    if (expr.deterministic) {
+      // For structural equivalent expressions, we need to pass in int type ordinals into
+      // split functions. If the number of ordinals is more than JVM function limit, we skip
+      // this expression.
+      // We calculate function parameter length by the number of ints plus `INPUT_ROW` plus
+      // a int type result array index.
+      val refs = expr.collect {
+        case _: BoundReference => Literal(0)
+      }
+      val parameterLength = CodeGenerator.calculateParamLength(refs) + 2
+      if (CodeGenerator.isValidParamLength(parameterLength)) {
+        val e: StructuralExpr = StructuralExpr(expr)
+        val f = structEquivalenceMap.get(e)
+        if (f.isDefined) {
+          addExpr(expr, f.get)
+        } else {
+          val exprMap = mutable.HashMap.empty[Expr, mutable.ArrayBuffer[Expression]]
+          addExpr(expr, exprMap)
+          structEquivalenceMap.put(e, exprMap)
+        }
+        true
+      } else {
+        false
+      }
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Checks if we skip add sub-expressions for given expression.
    */
-  def addExprTree(expr: Expression): Unit = {
-    val skip = expr.isInstanceOf[LeafExpression] ||
+  private def skipExpr(expr: Expression): Boolean = {
+    expr.isInstanceOf[LeafExpression] ||
       // `LambdaVariable` is usually used as a loop variable, which can't be evaluated ahead of the
       // loop. So we can't evaluate sub-expressions containing `LambdaVariable` at the beginning.
       expr.find(_.isInstanceOf[LambdaVariable]).isDefined
+  }
+
+
+  // There are some special expressions that we should not recurse into all of its children.
+  //   1. CodegenFallback: it's children will not be used to generate code (call eval() instead)
+  //   2. If: common subexpressions will always be evaluated at the beginning, but the true and
+  //          false expressions in `If` may not get accessed, according to the predicate
+  //          expression. We should only recurse into the predicate expression.
+  //   3. CaseWhen: like `If`, the children of `CaseWhen` only get accessed in a certain
+  //                condition. We should only recurse into the first condition expression as it
+  //                will always get accessed.
+  //   4. Coalesce: it's also a conditional expression, we should only recurse into the first
+  //                children, because others may not get accessed.
```

Comment:
nit: Do we need some words for `SortPrefix`?

Reply:
These words are copied from previous place. Let me add one for it.